### DISCUSSION TITLE
Got error when using k8s service discovery

### DISCUSSION THREAD

**Discussion by nhanh9d** [ThreeMammals/Ocelot#1948] [Show and tell]
## Actual Behavior
- After add solution from issue [977](https://github.com/ThreeMammals/Ocelot/issues/977), I still got error when calling service through k8s service discovery:
  - ![image](https://user-images.githubusercontent.com/41713343/117387151-80328a00-af12-11eb-86fd-ba07a74b6aab.png)
  - UnableToFindLoadBalancerError Message: unabe to find load balancer for /k8s/{url}|Options,Put,Get,Delete| exception is System.InvalidOperationException: KubeApiClient.CreateFromPodServiceAccount can only be called when running in a Kubernetes Pod (KUBERNETES_SERVICE_HOST and/or KUBERNETES_SERVICE_PORT environment variable is not defined)
- After I added 2 environment variables: KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT, I got another error: 
  - ![image](https://user-images.githubusercontent.com/41713343/117387428-1bc3fa80-af13-11eb-9359-a3dfa565cb38.png)
  - UnableToFindLoadBalancerError Message: unabe to find load balancer for /k8s/{url}|Options,Put,Get,Delete| exception is System.IO.DirectoryNotFoundException: Could not find a part of the path 'C:\var\run\secrets\kubernetes.io\serviceaccount\token'

## Steps to Reproduce the Problem

  1. Use Consul Service Discovery Provider
  2. Change to use K8s Service Discovery Provider
  3. Here is my config:
   ```json
"Routes": [
    //k8s test
    {
      "DownstreamPathTemplate": "/{url}",
      "DownstreamScheme": "http",
      "UpstreamPathTemplate": "/k8s/{url}",
      "UpstreamHttpMethod": [ "Options", "Put", "Get", "Delete" ],
      "ServiceName": "service-auth-admin-api"
    }
  ],
 "GlobalConfiguration": {
    "ServiceDiscoveryProvider": {
      //Specify the host of Service Discovery
      "Host": "172.20.40.141",
      //Specify the port of Service Discovery
      "Port": 6443,
      "Token": "MyToken",
      "Type": "kube",
      "Namespace": "admin-auth-server"
      ////1. Consul, means that Ocelot will get service information from consul per request
      ////2. PollConsul, means that Ocelot will poll Consul for latest service information
      //"Type": "Consul",
      ////Tells Ocelot how often to call Consul for changes in the service configuration
      //"PollingInterval": 60000
    },
    "RateLimitOptions": {
      "DisableRateLimitHeaders": false,
      "QuotaExceededMessage": "Your out!",
      "HttpStatusCode": 703
    }
  }
```

## Specifications

  - Version: 20H2
  - Platform: Window 10 Pro

**Comment by yoramda** [comment]
Hi,

1. Try to Add Type: `KubernetesServiceDiscoveryProvider` to 
```json
"GlobalConfiguration": {
 "ServiceDiscoveryProvider": {
   //Specify the host of Service Discovery
   "Host": "172.20.40.141",
   //Specify the port of Service Discovery
   "Port": 6443,
   "Token": "MyToken",
   "Type": "kube",
   "Namespace": "admin-auth-server"
   ////1. Consul, means that Ocelot will get service information from consul per request
   ////2. PollConsul, means that Ocelot will poll Consul for latest service information
   Type: "KubernetesServiceDiscoveryProvider",
   ////Tells Ocelot how often to call Consul for changes in the service configuration
   //"PollingInterval": 60000
 },
```
2. Add LoadBalncer 

Service Discovery
The following shows how to set up a Route using service discovery then select the LeastConnection load balancer.
```json
{
    "DownstreamPathTemplate": "/api/posts/{postId}",
    "DownstreamScheme": "https",
    "UpstreamPathTemplate": "/posts/{postId}",
    "UpstreamHttpMethod": [ "Put" ],
    "ServiceName": "product",
    "LoadBalancerOptions": {
        "Type": "LeastConnection"
    },
}
```
When this is set up Ocelot will lookup the downstream host and port from the service discover provider and load balance requests across any available services. If you add and remove services from the service discovery provider (consul) then Ocelot should respect this and stop calling services that have been removed and start calling services that have been added.

**Reply by raman-m** [reply]
> Try to Add Type: `KubernetesServiceDiscoveryProvider` to...

This is absolutely wrong!
The correct type of K8s provider is `Kube` ❗

**Comment by ghost** [comment]
@nhanh9d Hi An, have you ever successfully solved the issue? I am dealing with the same issue and also tried doing @yoramda 's suggestion but no luck.

**Comment by tomfanara** [comment]
Hello did anybody have any luck at this? I have tried reading the docs over and over and cannot get it to work. It works in Docker. I am having the same issue. Here is my routing file. Also do I need RBAC setup?
```json
{
  "Routes": [
    {
      "DownstreamPathTemplate": "/home",
      "DownstreamScheme": "http",
      "UpstreamPathTemplate": "/test",
      "ServiceName": "productshttpaggegator",
      "UpstreamHttpMethod": [ "Get" ]
    },
    {
      "DownstreamPathTemplate": "/api/products/{postId}",
      "DownstreamScheme": "http",
      "UpstreamPathTemplate": "/products/{postId}",
      "ServiceName": "productshttpaggegator",
      "UpstreamHttpMethod": [ "Get" ]
    }
  ],
        "GlobalConfiguration": {
          "ServiceDiscoveryProvider": {
      
            "Type": "KubernetesServiceDiscoveryProvider"

          }
        }

}
```

**Reply by raman-m** [reply]
`"Type": "KubernetesServiceDiscoveryProvider`

This type doesn't exist ❗

**Comment by raman-m** [comment]
Hello everyone! The issue #977 has been reopened today due to a critical inconsistency found in testing projects. It has been acknowledged following similar reports in issue https://github.com/ThreeMammals/Ocelot/issues/2178#issuecomment-2424755928.
The upcoming hotfix release will be version [23.3.6](https://github.com/ThreeMammals/Ocelot/releases/tag/23.3.6). Once the patch is released, this discussion will be concluded.

**Reply by raman-m** [reply]
- #2180
